File: src/FSharpAux/Regex.fs
namespace FSharpAux

open System.Text.RegularExpressions

/// Regex module for functional use of regular expressions
// Test regex under: http://regexstorm.net/tester
module Regex =

    //http://stackoverflow.com/questions/5684014/f-mapping-regular-expression-matches-with-active-patterns
    module Active = 
        /// Returns the first occurencing match of the pattern 
        let (|RegexMatchValue|_|) (regex:Regex) input =
            let m = regex.Match(input)    
            if m.Success then Some m.Value
            else None



        // Returns the matching group of the first occurence of the pattern 
        let (|FirstRegexGroup|_|) pattern input =
            let m = Regex.Match(input, pattern)
            if m.Success then Some(List.tail [ for g in m.Groups -> g ])
            else None

        /// Returns the matching group of all occurences of the pattern
        // For example see Mgf Parser
        let (|RegexGroups|_|) pattern input =
            let m = Regex.Matches(input, pattern)
            if m.Count > 0 then Some([ for m' in m -> m'.Groups ])
            else None


        /// Returns the matching values the first occurence of the pattern 
        let (|RegexValue|_|) pattern input =
            let m = Regex.Match(input, pattern)
            if m.Success then Some(List.tail [ for g in m.Groups -> g.Value ])
            else None

        /// Returns the matching values of all occurences of the pattern
        let (|RegexValues|_|) pattern input =
            let m = Regex.Matches(input, pattern)
            if m.Count > 0 then Some([ for g in m -> g.Value ])
            else None


        /// Returns the matching group values of all occurences of the pattern
        let (|RegexGroupValues|_|) pattern input =
            let m = Regex.Matches(input, pattern)
            if m.Count > 0 then Some([ for m' in m -> [for g in m'.Groups -> g.Value] ])
            else None



    /// Returns the first occurencing value of the match 
    let tryParseValue regexStr line = 
        let m = Regex.Match(line,regexStr)
        match m.Success with
        | true -> Some m.Value
        | false -> None

    /// Returns the group values of the first occurencing match
    let parse regexStr line = 
        let m = Regex.Match(line,regexStr)
        match m.Success with
        | true  -> (List.tail [ for g in m.Groups -> g.Value ])
        | false -> []

    /// Returns a seq of group values matching the pattern 
    let parseAll regexStr line = 
        let rec loop (m:Match) =
            seq {
            match m.Success with
            | true -> 
                yield (List.tail [ for g in m.Groups -> g.Value ])
                yield! loop (m.NextMatch())
            | false -> () }
        let m = Regex.Match(line,regexStr)
        loop m


    [<System.FlagsAttribute>]
    /// Provides enumerated values to use to set regular expressions options
    type RegexOptions =
        | None                     =   0
        | IgnoreCase               =   1
        | Multiline                =   2
        | ExplicitCapture          =   4
        | Compiled                 =   8
        | Singleline               =  16

        | IgnorePatternWhitespace  =  32
        | RightToLeft              =  64
        | ECMAScript               = 256
        | CultureInvariant         = 512

    /// Creates a regex 
    let createRegex (regexOptions:RegexOptions) pattern =
        new Regex(pattern, regexOptions |> box :?> System.Text.RegularExpressions.RegexOptions )

    

File: tests/FSharpAux.Mocha.Tests/fable/src/FSharpAux.Core/Regex.js
import { nextMatch, matches, create, match } from "../../fable_modules/fable-library.4.0.6/RegExp.js";
import { empty, tail } from "../../fable_modules/fable-library.4.0.6/List.js";
import { empty as empty_1, singleton, append, map, delay, toList } from "../../fable_modules/fable-library.4.0.6/Seq.js";

export function Active_$007CRegexMatchValue$007C_$007C(regex, input) {
    const m = match(regex, input);
    if (m != null) {
        return m[0];
    }
    else {
        return void 0;
    }
}

export function Active_$007CFirstRegexGroup$007C_$007C(pattern, input) {
    const m = match(create(pattern), input);
    if (m != null) {
        return tail(toList(delay(() => map((g) => g, m))));
    }
    else {
        return void 0;
    }
}

export function Active_$007CRegexGroups$007C_$007C(pattern, input) {
    const m = matches(create(pattern), input);
    if (m.length > 0) {
        return toList(delay(() => map((m$0027) => m$0027, m)));
    }
    else {
        return void 0;
    }
}

export function Active_$007CRegexValue$007C_$007C(pattern, input) {
    const m = match(create(pattern), input);
    if (m != null) {
        return tail(toList(delay(() => map((g) => (g || ""), m))));
    }
    else {
        return void 0;
    }
}

export function Active_$007CRegexValues$007C_$007C(pattern, input) {
    const m = matches(create(pattern), input);
    if (m.length > 0) {
        return toList(delay(() => map((g) => g[0], m)));
    }
    else {
        return void 0;
    }
}

export function Active_$007CRegexGroupValues$007C_$007C(pattern, input) {
    const m = matches(create(pattern), input);
    if (m.length > 0) {
        return toList(delay(() => map((m$0027) => toList(delay(() => map((g) => (g || ""), m$0027))), m)));
    }
    else {
        return void 0;
    }
}

export function tryParseValue(regexStr, line) {
    const m = match(create(regexStr), line);
    if (m != null) {
        return m[0];
    }
    else {
        return void 0;
    }
}

export function parse(regexStr, line) {
    const m = match(create(regexStr), line);
    if (m != null) {
        return tail(toList(delay(() => map((g) => (g || ""), m))));
    }
    else {
        return empty();
    }
}

export function parseAll(regexStr, line) {
    const loop = (m) => delay(() => {
        if (m != null) {
            return append(singleton(tail(toList(delay(() => map((g) => (g || ""), m))))), delay(() => loop(nextMatch(m))));
        }
        else {
            return empty_1();
        }
    });
    return loop(match(create(regexStr), line));
}

export function createRegex(regexOptions, pattern) {
    return create(pattern, regexOptions);
}

